### DISCUSSION TITLE
Explicit action for free text option

### DISCUSSION THREAD

**Discussion by j-manu** [josefarias/hotwire_combobox#15] [Ideas]
For free text option "ABC not found. Add ABC" will make the action explicit. Without users taking an explicit step, it's easy for typos to get added.

**Comment by josefarias** [comment]
Agree an empty (aka "blank") state would be nice. Would be great to support passing in both some text or a set of rendering options like in `:render_in`. Would welcome a PR for this.

That said, I think an empty state (i.e. "ABC not found") is separate from an explicit action for free text. Consider the following example:

![Screenshot shows "Aba" entered in the combobox, the listbox contains the suggestion "Alabama", because it makes a fuzzy match](https://github.com/josefarias/hotwire_combobox/assets/31393016/237aa8e5-a95d-4f11-b163-4feca758b50d)

Depending on your filtering algorithm and how fuzzy it is, free-text input could potentially make a fuzzy match with one of the legitimate options. We'd have to show the explicit button "Add ABC" in this scenario, too.

I'd like to avoid having many configurable options — that's how JS libraries infamously get out of hand and become difficult to use. So not sold on adding the explicit action for free-text options. But could be talked into it.

**Reply by josefarias** [reply]
> I'd like to avoid having many configurable options

By this I mean some people will prefer not having the explicit button, and some will prefer having it. I'd like to avoid supporting both via a configuration param, unless the implementation is trivial.

**Comment by josefarias** [comment]
Continuing the explicit "new option" button from https://github.com/josefarias/hotwire_combobox/issues/86, which I'm closing in favor of this.

**Comment by edwinv** [comment]
In our use cases we definitely would require an explicit "Add new" option. There can be matches in the existing options that are not the ones the user is looking for and hence needs to add a new one. So you can have several combinations:

- Full list with "add new" option
- Filtered list with "add new" option
- Empty list/state with "add new" option

The suggestion from @michoels to have a slot for this while rendering makes sense. We just always want to display this option, so there is no logic needed for displaying the option.

**Reply by Michoels** [reply]
I second this.
By using a slot, the complexity of rendering a button and handling interactions is wholly outsourced to the contents of the slot.

This library simply hosts an optional slot; that's all.

**Reply by Michoels** [reply]
(Optionally:
If we want to be more full-featured, we could pass the combobox state into the slot as an optional argument.
The slot could then choose to render differently in different situations.
This would add considerable complexity, so it may not be worth tackling at present.)

**Reply by josefarias** [reply]
> the complexity of rendering a button and handling interactions is wholly outsourced to the contents of the slot.

Not exactly. Because the combobox already supports free-text input and these two conceptualizations conflict with each other (because the option to enable free-text is called _name_when_new_).

I get the impression that both @edwinv and @Michoels are picturing a slot that links to a new flow when clicked. Is that correct?

In the current implementation, a new option is text you have actually entered that doesn't match any of the provided options exactly. That's useful in certain contexts, but seems like not yours.

In your proposal, a new option is a slot that links to a new flow (I think). That's useful in other contexts, but not the ones where one would need the current implementation.

If I had to choose between these two conceptualizations of what a "new option" is, I'd pick the current one. So I don't think we'll be implementing the slotted link unless we can come up with a new framing that suits both conceptualizations. Which could be a generic slot to insert _anything_ to the top of the options list. Not sure I love that, though.

For the time being, I'd suggest revisiting whether the new flow link needs to live inside the combobox. It could live somewhere in the page/form copy, for example.

**Reply by josefarias** [reply]
Actually, maybe I _could_ get behind a generic top-of-the-list slot 🤔 I'd welcome a proposal for this.

Would like to make sure we get the framing right and the code would need to be simple in order to make it worth it.

**Reply by josefarias** [reply]
In these conversations, we've gone in-and-out on the slot being generic or specific to new options. If someone were to make an implementation proposal for the generic slot, please do so in a new discussion or PR to keep this discussion limited to explicit "new options" features.

**Comment by josefarias** [comment]
Coming from https://github.com/josefarias/hotwire_combobox/discussions/14#discussioncomment-8873237 —

In that comment, the idea of adding instruction text to the combobox came up. The way I picture that is by injecting it to the top of the listbox.

That "injection" made me think of @Michoels's proposal of a generic slot for inserting anything to the top of the listbox.

I wanna keep momentum moving forward so I'm inclined to make a decision here — it's gonna be a "no" for an explicit action for free text. And an optimistic "maybe" for the generic slot, which I've added to my notes to keep track of.

I still kind of like my placebo button idea. If we ever implement an explicit freetext button I think it would be that way. But I worry about how that interacts with the generic slot at the top. Need to do some more thinking around this. And because I'm not convinced enough to say "yes" right now, I'll say "no" and let the idea gestate until the correct framing comes to light.

**Reply by Michoels** [reply]
Sounds good.
Instruction text would be a good use of a generic slot.

Should I open a new Discussion about it?

**Reply by josefarias** [reply]
If you'd like to and have time, that'd be great @Michoels! Thank you!

It's on my notes but a discussion would be a good place for folks to share what they'd like in the feature. And maybe someone can get to it before I can.

**Comment by gnapse** [comment]
An alternative way to provide some "solution" to this is if we could be able to be notified somehow that a new option is selected. I arrived here because what I wanted was a way to change something around the combobox that would alert the user that they selected to create a new option. For instance, adding a warning or hint below the combobox. That would be enough for me. And would alert users if they didn't mean to add a new option.

Is something like that possible?

**Reply by Michoels** [reply]
You can detect when an option is selected by listening for the `hw-combobox:selection` event.
See the full syntax [here](https://github.com/josefarias/hotwire_combobox/issues/181#issuecomment-2536157712).

**Reply by gnapse** [reply]
Thanks for the information!

Is there a way to discover these features? I checked, and this is not documented in this repo's README, and it's also not mentioned on the website https://hotwirecombobox.com (is it?)

I now wonder what other similar useful APIs this component has that are not mentioned in the documentation.

cc @josefarias

**Reply by Michoels** [reply]
In the [_Towards 1.0.0_ discussion](https://github.com/josefarias/hotwire_combobox/discussions/200), @josefarias noted that:

> Specifically, I've held off on documenting (and thus finalizing) multiselect, custom DOM events, and custom attributes just so we could live with them for a while and make any breaking decisions before people adopt them.
